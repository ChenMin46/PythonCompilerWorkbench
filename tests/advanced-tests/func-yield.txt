def foo():
    yield

def foo():
    yield 42

def foo():
    yield True

def foo():
    yield a + b

def foo():
    yield foo(n)

def fact(n):
    yield n * fact(n-1)
