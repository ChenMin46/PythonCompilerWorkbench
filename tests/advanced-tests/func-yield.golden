def foo():
    yield

def foo():
    yield 42

def foo():
    yield True

def foo():
    yield a + b

def foo():
    yield foo(n)

def fact(n):
    yield n * fact(n-1)
===
Module
  body: [
    FunctionDef >> 'def foo()'
      name: 'foo'
      args: arguments
        args: [
        ]
        vararg: None
        kwarg: None
        defaults: [
        ]
      body: [
        Expr
          value: Yield >> 'yield'
            value: None
      ]
      decorator_list: [
      ]
    FunctionDef >> 'def foo()'
      name: 'foo'
      args: arguments
        args: [
        ]
        vararg: None
        kwarg: None
        defaults: [
        ]
      body: [
        Expr
          value: Yield >> 'yield 42'
            value: Num >> '42'
              n: 42
      ]
      decorator_list: [
      ]
    FunctionDef >> 'def foo()'
      name: 'foo'
      args: arguments
        args: [
        ]
        vararg: None
        kwarg: None
        defaults: [
        ]
      body: [
        Expr
          value: Yield >> 'yield True'
            value: Name >> 'True'
              id: 'True'
              ctx: Load
      ]
      decorator_list: [
      ]
    FunctionDef >> 'def foo()'
      name: 'foo'
      args: arguments
        args: [
        ]
        vararg: None
        kwarg: None
        defaults: [
        ]
      body: [
        Expr
          value: Yield >> 'yield a + b'
            value: BinOp >> 'a + b'
              left: Name >> 'a'
                id: 'a'
                ctx: Load
              op: Add
              right: Name >> 'b'
                id: 'b'
                ctx: Load
      ]
      decorator_list: [
      ]
    FunctionDef >> 'def foo()'
      name: 'foo'
      args: arguments
        args: [
        ]
        vararg: None
        kwarg: None
        defaults: [
        ]
      body: [
        Expr
          value: Yield >> 'yield foo(n)'
            value: Call >> 'foo(n)'
              func: Name >> 'foo'
                id: 'foo'
                ctx: Load
              args: [
                Name >> 'n'
                  id: 'n'
                  ctx: Load
              ]
              keywords: [
              ]
              starargs: None
              kwargs: None
      ]
      decorator_list: [
      ]
    FunctionDef >> 'def fact(n)'
      name: 'fact'
      args: arguments
        args: [
          Name >> 'n'
            id: 'n'
            ctx: Param
        ]
        vararg: None
        kwarg: None
        defaults: [
        ]
      body: [
        Expr
          value: Yield >> 'yield n * fact(n-1)'
            value: BinOp >> 'n * fact(n-1)'
              left: Name >> 'n'
                id: 'n'
                ctx: Load
              op: Mult
              right: Call >> 'fact(n-1)'
                func: Name >> 'fact'
                  id: 'fact'
                  ctx: Load
                args: [
                  BinOp >> 'n-1'
                    left: Name >> 'n'
                      id: 'n'
                      ctx: Load
                    op: Sub
                    right: Num >> '1'
                      n: 1
                ]
                keywords: [
                ]
                starargs: None
                kwargs: None
      ]
      decorator_list: [
      ]
  ]
--- JSON ---
{
  "contents": [
    {
      "contents": [
        {
          "contents": [
            "def foo()"
          ], 
          "id": "id_0", 
          "name": "FunctionDef"
        }, 
        {
          "id": "tid_1", 
          "name": "token", 
          "value": ":"
        }, 
        "\n", 
        "    ", 
        {
          "contents": [
            {
              "contents": [
                {
                  "id": "tid_2", 
                  "name": "token", 
                  "value": "yield"
                }
              ], 
              "id": "id_0", 
              "name": "Yield"
            }
          ], 
          "id": "id_0", 
          "name": "Expr"
        }, 
        "\n\n", 
        {
          "contents": [
            "def foo()"
          ], 
          "id": "id_0", 
          "name": "FunctionDef"
        }, 
        {
          "id": "tid_3", 
          "name": "token", 
          "value": ":"
        }, 
        "\n", 
        "    ", 
        {
          "contents": [
            {
              "contents": [
                {
                  "id": "tid_4", 
                  "name": "token", 
                  "value": "yield"
                }, 
                " ", 
                {
                  "contents": [
                    "42"
                  ], 
                  "id": "id_0", 
                  "name": "Num"
                }
              ], 
              "id": "id_0", 
              "name": "Yield"
            }
          ], 
          "id": "id_0", 
          "name": "Expr"
        }, 
        "\n\n", 
        {
          "contents": [
            "def foo()"
          ], 
          "id": "id_0", 
          "name": "FunctionDef"
        }, 
        {
          "id": "tid_5", 
          "name": "token", 
          "value": ":"
        }, 
        "\n", 
        "    ", 
        {
          "contents": [
            {
              "contents": [
                {
                  "id": "tid_6", 
                  "name": "token", 
                  "value": "yield"
                }, 
                " ", 
                {
                  "contents": [
                    "True"
                  ], 
                  "id": "id_0", 
                  "name": "Name"
                }
              ], 
              "id": "id_0", 
              "name": "Yield"
            }
          ], 
          "id": "id_0", 
          "name": "Expr"
        }, 
        "\n\n", 
        {
          "contents": [
            "def foo()"
          ], 
          "id": "id_0", 
          "name": "FunctionDef"
        }, 
        {
          "id": "tid_7", 
          "name": "token", 
          "value": ":"
        }, 
        "\n", 
        "    ", 
        {
          "contents": [
            {
              "contents": [
                {
                  "id": "tid_8", 
                  "name": "token", 
                  "value": "yield"
                }, 
                " ", 
                {
                  "contents": [
                    {
                      "contents": [
                        "a"
                      ], 
                      "id": "id_0", 
                      "name": "Name"
                    }, 
                    " ", 
                    {
                      "id": "tid_9", 
                      "name": "token", 
                      "value": "+"
                    }, 
                    " ", 
                    {
                      "contents": [
                        "b"
                      ], 
                      "id": "id_0", 
                      "name": "Name"
                    }
                  ], 
                  "id": "id_0", 
                  "name": "BinOp"
                }
              ], 
              "id": "id_0", 
              "name": "Yield"
            }
          ], 
          "id": "id_0", 
          "name": "Expr"
        }, 
        "\n\n", 
        {
          "contents": [
            "def foo()"
          ], 
          "id": "id_0", 
          "name": "FunctionDef"
        }, 
        {
          "id": "tid_10", 
          "name": "token", 
          "value": ":"
        }, 
        "\n", 
        "    ", 
        {
          "contents": [
            {
              "contents": [
                {
                  "id": "tid_11", 
                  "name": "token", 
                  "value": "yield"
                }, 
                " ", 
                {
                  "contents": [
                    {
                      "contents": [
                        "foo"
                      ], 
                      "id": "id_0", 
                      "name": "Name"
                    }, 
                    {
                      "id": "tid_12", 
                      "name": "token", 
                      "value": "("
                    }, 
                    {
                      "contents": [
                        "n"
                      ], 
                      "id": "id_0", 
                      "name": "Name"
                    }, 
                    {
                      "id": "tid_13", 
                      "name": "token", 
                      "value": ")"
                    }
                  ], 
                  "id": "id_0", 
                  "name": "Call"
                }
              ], 
              "id": "id_0", 
              "name": "Yield"
            }
          ], 
          "id": "id_0", 
          "name": "Expr"
        }, 
        "\n\n", 
        {
          "contents": [
            "def fact(", 
            {
              "contents": [
                {
                  "contents": [
                    "n"
                  ], 
                  "id": "id_0", 
                  "name": "Name"
                }
              ], 
              "id": "id_0", 
              "name": "arguments"
            }, 
            {
              "id": "tid_14", 
              "name": "token", 
              "value": ")"
            }
          ], 
          "id": "id_0", 
          "name": "FunctionDef"
        }, 
        {
          "id": "tid_15", 
          "name": "token", 
          "value": ":"
        }, 
        "\n", 
        "    ", 
        {
          "contents": [
            {
              "contents": [
                {
                  "id": "tid_16", 
                  "name": "token", 
                  "value": "yield"
                }, 
                " ", 
                {
                  "contents": [
                    {
                      "contents": [
                        "n"
                      ], 
                      "id": "id_0", 
                      "name": "Name"
                    }, 
                    " ", 
                    {
                      "id": "tid_17", 
                      "name": "token", 
                      "value": "*"
                    }, 
                    " ", 
                    {
                      "contents": [
                        {
                          "contents": [
                            "fact"
                          ], 
                          "id": "id_0", 
                          "name": "Name"
                        }, 
                        {
                          "id": "tid_18", 
                          "name": "token", 
                          "value": "("
                        }, 
                        {
                          "contents": [
                            {
                              "contents": [
                                "n"
                              ], 
                              "id": "id_0", 
                              "name": "Name"
                            }, 
                            {
                              "id": "tid_19", 
                              "name": "token", 
                              "value": "-"
                            }, 
                            {
                              "contents": [
                                "1"
                              ], 
                              "id": "id_0", 
                              "name": "Num"
                            }
                          ], 
                          "id": "id_0", 
                          "name": "BinOp"
                        }, 
                        {
                          "id": "tid_20", 
                          "name": "token", 
                          "value": ")"
                        }
                      ], 
                      "id": "id_0", 
                      "name": "Call"
                    }
                  ], 
                  "id": "id_0", 
                  "name": "BinOp"
                }
              ], 
              "id": "id_0", 
              "name": "Yield"
            }
          ], 
          "id": "id_0", 
          "name": "Expr"
        }
      ], 
      "id": "id_0", 
      "name": "Module"
    }, 
    "\n"
  ], 
  "id": "id_ROOT", 
  "name": "ROOT_NODE"
}
